service/storage_service.cc

```diff
@@ -840,17 +896,75 @@ void storage_service::mark_existing_views_as_built() {
     }).get();
 }
 
+// Run inside async
+db_clock::time_point storage_service::make_new_cdc_generation() {
+    // We are prepared to bootstrap or join the ring: we know what our tokens will be.
+    assert(!_bootstrap_tokens.empty());
+
+    // After we pick a generation timestamp, we start gossiping it, and we stick with it.
+    // We don't do any other generation switches (unless we crash before complecting bootstrap).
+    assert(!_streams_ts);
+
+    auto gen = cdc::generate_cluster_topology_description(
+            cdc::db_context::builder(get_local_storage_proxy())
+                .with_token_metadata(_token_metadata)
+                .with_snitch(locator::i_endpoint_snitch::get_local_snitch_ptr())
+                .with_partitioner(dht::global_partitioner())
+                .build(),
+            _bootstrap_tokens);
+
+    // Begin the race.
+    auto ts = db_clock::now() + 2 * get_ring_delay();
+    _sys_dist_ks.local().insert_cdc_cluster_topology_description(ts, std::move(gen)).get();
+
+    return ts;
+}
+
+// Runs inside seastar::async context.
+bool storage_service::should_propose_first_cdc_generation() {
+    auto my_host_id = _gossiper.get_host_id(get_broadcast_address());
+    for (const auto& ep: _gossiper.get_endpoint_states()) {
+        if (my_host_id < _gossiper.get_host_id(ep.first)) {
+            return false;
+        }
+    }
+    return true;
+}
+
 // Runs inside seastar::async context
 void storage_service::bootstrap() {
     _is_bootstrap_mode = true;
     if (!db().local().is_replacing()) {
         // Wait until we know tokens of existing node before announcing join status.
         _gossiper.wait_for_range_setup().get();
-        // if not an existing token then bootstrap
-        _gossiper.add_local_application_state({
-            { gms::application_state::TOKENS, value_factory.tokens(_bootstrap_tokens) },
-            { gms::application_state::STATUS, value_factory.bootstrapping(_bootstrap_tokens) },
-        }).get();
+
+        // Even if we reached this point before but crashed, we will make a new CDC generation.
+        // It doesn't hurt: other nodes will (potentially) just do more generation switches.
+        // We do this because with this new attempt at bootstrapping we picked a different set of tokens.
+
+        if (db().local().get_config().check_experimental(db::experimental_features_t::CDC)) {
+            _streams_ts = make_new_cdc_generation();
+        } else {
+            // We should not be able to join the cluster if other nodes support CDC but we don't.
+            // The check should have been made somewhere in prepare_to_join (`check_knows_remote_features`).
+            assert(!cluster_supports_cdc());
+        }
+
+        if (_streams_ts) {
```

Comment:
If we change to add cdc stream timestamp unconditionally, this if branch is eliminated.

Reply:
Done